packages/core/crates/core-crypto/src/routing.rs

```diff
@@ -0,0 +1,381 @@
+    let last_hop = 1 + additional_data_last_hop_len;
+
+    last_hop + (max_hops - 1) * per_hop
+}
+
+fn generate_filler(
+    max_hops: usize,
+    routing_info_len: usize,
+    routing_info_last_hop_len: usize,
+    secrets: &[&[u8]],
+) -> Box<[u8]> {
+    if secrets.len() < 2 {
+        return vec![].into_boxed_slice();
+    }
+
+    assert!(max_hops >= secrets.len(), "too few hops");
+    assert!(routing_info_len > 0, "invalid routing info length");
+
+    let header_len = routing_info_last_hop_len + (max_hops - 1) * routing_info_len;
+    let padding_len = (max_hops - secrets.len()) * routing_info_len;
+
+    let mut ret = vec![0u8; header_len - padding_len - routing_info_last_hop_len];
+
+    let mut length = routing_info_len;
+    let mut start = header_len;
+
+    for &secret in secrets.iter().take(secrets.len() - 1) {
+        let prg = PRG::from_parameters(PRGParameters::new(secret));
+
+        let digest = prg.digest(start, header_len + routing_info_len);
+        xor_inplace(&mut ret[0..length], digest.as_ref());
+
+        length += routing_info_len;
+        start -= routing_info_len;
+    }
+
+    ret.into_boxed_slice()
+}
+
+/// Carries routing information for the mixnet packet.
+pub struct RoutingInfo {
+    pub routing_information: Box<[u8]>,
+    pub mac: [u8; SimpleMac::SIZE],
+}
+
+impl Default for RoutingInfo {
+    fn default() -> Self {
+        Self {
+            routing_information: Box::default(),
+            mac: [0u8; SimpleMac::SIZE],
+        }
+    }
+}
+
+impl RoutingInfo {
+    /// Creates the routing information of the mixnet packet
+    /// # Arguments
+    /// * `max_hops` maximal number of hops
+    /// * `path` IDs of the nodes along the path
+    /// * `secrets` shared secrets with the nodes along the path
+    /// * `additional_data_relayer_len` length of each additional data for all relayers
+    /// * `additional_data_relayer` additional data for each relayer
+    /// * `additional_data_last_hop` additional data for the final recipient
+    pub fn new(
+        max_hops: usize,
+        path: &[PublicKey],
+        secrets: &[&[u8]],
+        additional_data_relayer_len: usize,
+        additional_data_relayer: &[&[u8]],
+        additional_data_last_hop: Option<&[u8]>,
+    ) -> Self {
+        assert!(
+            secrets.len() <= max_hops && !secrets.is_empty(),
+            "invalid number of secrets given"
+        );
+        assert!(
+            secrets.iter().all(|s| s.len() == SECRET_KEY_LENGTH),
+            "invalid secret length"
+        );
+        assert!(
+            additional_data_relayer
+                .iter()
+                .all(|r| r.len() == additional_data_relayer_len),
+            "invalid relayer data length"
+        );
+        assert!(
+            additional_data_last_hop.is_none() || !additional_data_last_hop.unwrap().is_empty(),
+            "invalid additional data for last hop"
+        );
```

Comment:
So that's the question: Is this serialization or deserialization of the packet? If it's serialization, it should likely use native types that would do the asserts on object creation, if it's a deserialization, asserts could likely be reworked as an iterator failure over a byte sequence, in which case it would fail JIT once encountering an invalid format.

Reply:
OK, then we should probably take arguments as types... construct the packet as a type... and a finalized, successful construction convert into a binary form.

Reply:
Yes I agree. More type-induced safety with new types will be introduced in the subsequent curve switch PR.